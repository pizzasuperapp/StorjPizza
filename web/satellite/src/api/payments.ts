// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { ErrorUnauthorized } from '@/api/errors/ErrorUnauthorized';
import {
    AccountBalance,
    Coupon,
    CreditCard,
    PaymentsApi,
    PaymentsHistoryItem,
    ProjectUsageAndCharges,
    TokenDeposit
} from '@/types/payments';
import { HttpClient } from '@/utils/httpClient';
import { Time } from '@/utils/time';
import { ErrorTooManyRequests } from './errors/ErrorTooManyRequests';

/**
 * PaymentsHttpApi is a http implementation of Payments API.
 * Exposes all payments-related functionality
 */
export class PaymentsHttpApi implements PaymentsApi {
    private readonly client: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/payments';

    /**
     * Get account balance.
     *
     * @returns balance in cents
     * @throws Error
     */
    public async getBalance(): Promise<AccountBalance> {
        const path = `${this.ROOT_PATH}/account/balance`;
        const response = await this.client.get(path);

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }

            throw new Error('Can not get account balance');
        }

        const balance = await response.json();
        if (balance) {
            return new AccountBalance(balance.freeCredits, balance.coins);
        }

        return new AccountBalance();
    }

    /**
     * Try to set up a payment account.
     *
     * @throws Error
     */
    public async setupAccount(): Promise<string> {
        const path = `${this.ROOT_PATH}/account`;
        const response = await this.client.post(path, null);
        const couponType = await response.json();

        if (response.ok) {
            return couponType;
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not setup account');
    }

    /**
     * projectsUsageAndCharges returns usage and how much money current user will be charged for each project which he owns.
     */
    public async projectsUsageAndCharges(start: Date, end: Date): Promise<ProjectUsageAndCharges[]> {
        const since = Time.toUnixTimestamp(start).toString();
        const before = Time.toUnixTimestamp(end).toString();
        const path = `${this.ROOT_PATH}/account/charges?from=${since}&to=${before}`;
        const response = await this.client.get(path);

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }

            throw new Error('can not get projects charges');
        }

        const charges = await response.json();
        if (charges) {
            return charges.map(charge =>
                new ProjectUsageAndCharges(
                    new Date(charge.since),
                    new Date(charge.before),
                    charge.egress,
                    charge.storage,
                    charge.segmentCount,
                    charge.projectId,
                    charge.storagePrice,
                    charge.egressPrice,
                    charge.segmentPrice,
                ),
            );
        }

        return [];
    }

    /**
     * Add credit card.
     *
     * @param token - stripe token used to add a credit card as a payment method
     * @throws Error
     */
    public async addCreditCard(token: string): Promise<void> {
        const path = `${this.ROOT_PATH}/cards`;
        const response = await this.client.post(path, token);

        if (response.ok) {
            return;
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not add credit card');
    }

    /**
     * Detach credit card from payment account.
     *
     * @param cardId
     * @throws Error
     */
    public async removeCreditCard(cardId: string): Promise<void> {
        const path = `${this.ROOT_PATH}/cards/${cardId}`;
        const response = await this.client.delete(path);

        if (response.ok) {
            return;
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not remove credit card');
    }

    /**
     * Get list of user`s credit cards.
     *
     * @returns list of credit cards
     * @throws Error
     */
    public async listCreditCards(): Promise<CreditCard[]> {
        const path = `${this.ROOT_PATH}/cards`;
        const response = await this.client.get(path);

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }
            throw new Error('can not list credit cards');
        }

        const creditCards = await response.json();

        if (creditCards) {
            return creditCards.map(card => new CreditCard(card.id, card.expMonth, card.expYear, card.brand, card.last4, card.isDefault));
        }

        return [];
    }

    /**
     * Make credit card default.
     *
     * @param cardId
     * @throws Error
     */
    public async makeCreditCardDefault(cardId: string): Promise<void> {
        const path = `${this.ROOT_PATH}/cards`;
        const response = await this.client.patch(path, cardId);

        if (response.ok) {
            return;
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not make credit card default');
    }

    /**
     * Returns a list of invoices, transactions and all others payments history items for payment account.
     *
     * @returns list of payments history items
     * @throws Error
     */
    public async paymentsHistory(): Promise<PaymentsHistoryItem[]> {
        const path = `${this.ROOT_PATH}/billing-history`;
        const response = await this.client.get(path);

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }
            throw new Error('can not list billing history');
        }

        const paymentsHistoryItems = await response.json();
        if (paymentsHistoryItems) {
            return paymentsHistoryItems.map(item =>
                new PaymentsHistoryItem(
                    item.id,
                    item.description,
                    item.amount,
                    item.received,
                    item.status,
                    item.link,
                    new Date(item.start),
                    new Date(item.end),
                    item.type,
                    item.remaining,
                ),
            );
        }

        return [];
    }

    /**
     * makeTokenDeposit process coin payments.
     *
     * @param amount
     * @throws Error
     */
    public async makeTokenDeposit(amount: number): Promise<TokenDeposit> {
        const path = `${this.ROOT_PATH}/tokens/deposit`;
        const response = await this.client.post(path, JSON.stringify({ amount }));

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }

            throw new Error('can not process coin payment');
        }

        const result = await response.json();

        return new TokenDeposit(result.amount, result.address, result.link);
    }

    /**
     * applyCouponCode applies a coupon code.
     *
     * @param couponCode
     * @throws Error
     */
    public async applyCouponCode(couponCode: string): Promise<Coupon> {
        const path = `${this.ROOT_PATH}/coupon/apply`;
        const response = await this.client.patch(path, couponCode);
        const errMsg = `Could not apply coupon code "${couponCode}"`;

        if (response.ok) {
            const coupon = await response.json();

            if (!coupon) {
                throw new Error(errMsg);
            }

            return new Coupon(
                coupon.id,
                coupon.promoCode,
                coupon.name,
                coupon.amountOff,
                coupon.percentOff,
                new Date(coupon.addedAt),
                coupon.expiresAt ? new Date(coupon.expiresAt) : null,
                coupon.duration
            );
        }

        switch (response.status) {
        case 429:
            throw new ErrorTooManyRequests('You\'ve exceeded limit of attempts, try again in 5 minutes');
        case 401:
            throw new ErrorUnauthorized(errMsg);
        default:
            throw new Error(errMsg);
        }
    }

    /**
     * getCoupon returns the coupon applied to the user.
     *
     * @throws Error
     */
    public async getCoupon(): Promise<Coupon | null> {
        const path = `${this.ROOT_PATH}/coupon`;
        const response = await this.client.get(path);
        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }

            throw new Error('cannot retrieve coupon');
        }

        const coupon = await response.json();

        if (!coupon) {
            return null;
        }

        return new Coupon(
            coupon.id,
            coupon.promoCode,
            coupon.name,
            coupon.amountOff,
            coupon.percentOff,
            new Date(coupon.addedAt),
            coupon.expiresAt ? new Date(coupon.expiresAt) : null,
            coupon.duration
        );
    }
}
