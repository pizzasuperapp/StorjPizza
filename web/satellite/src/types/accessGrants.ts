// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { SortDirection } from '@/types/common';

export type OnHeaderClickCallback = (sortBy: AccessGrantsOrderBy, sortDirection: SortDirection) => Promise<void>;

/**
 * AccessGrantsWorker provides access to the WASM module.
 */
export interface AccessGrantsWorkerFactory {
    // TODO: this should be converted to a proper interface.
    create(): Worker;
}

/**
 * Exposes all access grants-related functionality.
 */
export interface AccessGrantsApi {
    /**
     * Fetch accessGrants
     *
     * @returns AccessGrantsPage
     * @throws Error
     */
    get(projectId: string, cursor: AccessGrantCursor): Promise<AccessGrantsPage>;

    /**
     * Create new accessGrant
     *
     * @returns AccessGrant
     * @throws Error
     */
    create(projectId: string, name: string): Promise<AccessGrant>;

    /**
     * Delete existing access grant
     *
     * @returns null
     * @throws Error
     */
    delete(ids: string[]): Promise<void>;

    /**
     * Delete existing access grant by name and project id
     *
     * @returns null
     * @throws Error
     */
    deleteByNameAndProjectID(name: string, projectID: string): Promise<void>;

    /**
     * Get gateway credentials using access grant
     *
     * @returns EdgeCredentials
     * @throws Error
     */
    getGatewayCredentials(accessGrant: string, optionalURL?: string, isPublic?: boolean): Promise<EdgeCredentials>;
}

/**
 * Holds access grants sorting parameters.
 */
export enum AccessGrantsOrderBy {
    NAME = 1,
    CREATED_AT,
}

/**
 * AccessGrantCursor is a type, used to describe paged access grants list.
 */
export class AccessGrantCursor {
    public constructor(
        public search: string = '',
        public limit: number = 6,
        public page: number = 1,
        public order: AccessGrantsOrderBy = AccessGrantsOrderBy.NAME,
        public orderDirection: SortDirection = SortDirection.ASCENDING,
    ) {}
}

/**
 * AccessGrantsPage is a type, used to describe paged access grants list.
 */
export class AccessGrantsPage {
    public constructor(
        public accessGrants: AccessGrant[] = [],
        public search: string = '',
        public order: AccessGrantsOrderBy = AccessGrantsOrderBy.NAME,
        public orderDirection: SortDirection = SortDirection.ASCENDING,
        public limit: number = 6,
        public pageCount: number = 0,
        public currentPage: number = 1,
        public totalCount: number = 0,
    ) {}
}

/**
 * AccessGrant class holds info for Access Grant entity.
 */
export class AccessGrant {
    public isSelected: boolean;

    constructor(
        public id: string = '',
        public name: string = '',
        public createdAt: Date = new Date(),
        public secret: string = '',
    ) {
        this.isSelected = false;
    }

    /**
     * Returns created date as a local string.
     */
    public localDate(): string {
        return this.createdAt.toLocaleString('en-US', {timeZone: 'UTC', timeZoneName: 'short'});
    }
}

/**
 * DurationPermission class holds info for access grant's duration permission.
 */
export class DurationPermission {
    constructor(
        public notBefore: Date | null = null,
        public notAfter: Date | null = null,
    ) {}
}

/**
 * EdgeCredentials class holds info for edge credentials generated from access grant.
 */
export class EdgeCredentials {
    constructor(
        public id: string = '',
        public createdAt: Date = new Date(),
        public accessKeyId: string = '',
        public secretKey: string = '',
        public endpoint: string = '',
    ) {}
}
