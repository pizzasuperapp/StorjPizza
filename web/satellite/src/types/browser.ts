// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Exposes all properties and methods present and available in the file/browser objects in Browser.
 */
export interface BrowserFile extends File {
  Key: string;
  LastModified: Date;
  Size: number;
  type: string;
}
