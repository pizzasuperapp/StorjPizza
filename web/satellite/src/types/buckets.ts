// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Exposes all bucket-related functionality.
 */
export interface BucketsApi {
    /**
     * Fetch buckets
     *
     * @returns BucketPage
     * @throws Error
     */
    get(projectId: string, before: Date, cursor: BucketCursor): Promise<BucketPage>;

    /**
     * Fetch all bucket names
     *
     * @returns string[]
     * @throws Error
     */
    getAllBucketNames(projectId: string): Promise<string[]>;
}

/**
 * Bucket class holds info for Bucket entity.
 */
export class Bucket {
    public constructor(
        public name: string = '',
        public storage: number = 0,
        public egress: number = 0,
        public objectCount: number = 0,
        public segmentCount: number = 0,
        public since: Date = new Date(),
        public before: Date = new Date(),
    ) { }
}

/**
 * BucketPage class holds bucket total usages and flag whether more usages available.
 */
export class BucketPage {
    public constructor(
        public buckets: Bucket[] = [],
        public search: string = '',
        public limit: number = 0,
        public offset: number = 0,
        public pageCount: number = 0,
        public currentPage: number = 0,
        public totalCount: number = 0,
    ) { }
}

/**
 * BucketCursor class holds cursor for bucket name and limit.
 */
export class BucketCursor {
    public constructor(
        public search: string = '',
        public limit: number = 0,
        public page: number = 0,
    ) { }
}
