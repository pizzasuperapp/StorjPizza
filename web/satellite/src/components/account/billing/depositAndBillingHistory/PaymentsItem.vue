// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="container">
        <PaymentsHistoryItemDate
            class="container__item date"
            :start="billingItem.start"
            :expiration="billingItem.end"
            :type="billingItem.type"
            :status="billingItem.status"
        />
        <p class="container__item description">{{ billingItem.description }}</p>
        <p class="container__item status">{{ billingItem.formattedStatus }}</p>
        <p class="container__item amount">
            <b>
                {{ billingItem.quantity.currency }}
                <span v-if="billingItem.type === 1">
                    {{ billingItem.quantity.received.toFixed(2) }}
                </span>
                <span v-else>
                    {{ billingItem.quantity.total.toFixed(2) }}
                </span>
            </b>
            <span v-if="billingItem.type === 1">
                of <b>{{ billingItem.quantity.total.toFixed(2) }}</b>
            </span>
        </p>
        <p class="container__item download">
            <a v-if="billingItem.link" class="download-link" target="_blank" :href="billingItem.link">{{ billingItem.label }}</a>
        </p>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import PaymentsHistoryItemDate from '@/components/account/billing/depositAndBillingHistory/PaymentsHistoryItemDate.vue';

import { PaymentsHistoryItem } from '@/types/payments';

// @vue/component
@Component({
    components: {
        PaymentsHistoryItemDate,
    },
})
export default class PaymentsItem extends Vue {
    @Prop({default: () => new PaymentsHistoryItem()})
    private readonly billingItem: PaymentsHistoryItem;
}
</script>

<style scoped lang="scss">
    .download-link {
        color: #2683ff;
        font-family: 'font_bold', sans-serif;

        &:hover {
            color: #0059d0;
        }
    }

    .container {
        display: flex;
        align-items: center;
        width: 100%;
        border-top: 1px solid #c7cdd2;

        &__item {
            min-width: 20%;
            font-family: 'font_medium', sans-serif;
            font-size: 16px;
            text-align: left;
            color: #768394;
            margin: 30px 0;
        }
    }

    .date {
        font-family: 'font_bold', sans-serif;
        margin: 0;
    }

    .description {
        min-width: 31%;
    }

    .status {
        min-width: 17%;
    }

    .amount {
        min-width: 22%;
        margin: 0;
    }

    .download {
        margin: 0;
        text-align: right;
        width: 10%;
        min-width: 10%;
    }

    .row {
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: flex-start;
        width: 175px;
    }
</style>
