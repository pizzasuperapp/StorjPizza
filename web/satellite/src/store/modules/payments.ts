// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    AccountBalance,
    Coupon,
    CreditCard,
    DateRange,
    PaymentsApi,
    PaymentsHistoryItem,
    PaymentsHistoryItemStatus,
    PaymentsHistoryItemType,
    ProjectUsageAndCharges,
    TokenDeposit,
} from '@/types/payments';
import {StoreModule} from "@/types/store";

export const PAYMENTS_MUTATIONS = {
    SET_BALANCE: 'SET_BALANCE',
    SET_CREDIT_CARDS: 'SET_CREDIT_CARDS',
    SET_DATE: 'SET_DATE',
    CLEAR: 'CLEAR_PAYMENT_INFO',
    UPDATE_CARDS_SELECTION: 'UPDATE_CARDS_SELECTION',
    UPDATE_CARDS_DEFAULT: 'UPDATE_CARDS_DEFAULT',
    SET_PAYMENTS_HISTORY: 'SET_PAYMENTS_HISTORY',
    SET_PROJECT_USAGE_AND_CHARGES: 'SET_PROJECT_USAGE_AND_CHARGES',
    SET_CURRENT_ROLLUP_PRICE: 'SET_CURRENT_ROLLUP_PRICE',
    SET_PREVIOUS_ROLLUP_PRICE: 'SET_PREVIOUS_ROLLUP_PRICE',
    SET_PRICE_SUMMARY: 'SET_PRICE_SUMMARY',
    SET_PRICE_SUMMARY_FOR_SELECTED_PROJECT: 'SET_PRICE_SUMMARY_FOR_SELECTED_PROJECT',
    SET_COUPON: 'SET_COUPON',
};

export const PAYMENTS_ACTIONS = {
    GET_BALANCE: 'getBalance',
    SETUP_ACCOUNT: 'setupAccount',
    GET_CREDIT_CARDS: 'getCreditCards',
    ADD_CREDIT_CARD: 'addCreditCard',
    CLEAR_PAYMENT_INFO: 'clearPaymentInfo',
    TOGGLE_CARD_SELECTION: 'toggleCardSelection',
    CLEAR_CARDS_SELECTION: 'clearCardsSelection',
    MAKE_CARD_DEFAULT: 'makeCardDefault',
    REMOVE_CARD: 'removeCard',
    GET_PAYMENTS_HISTORY: 'getPaymentsHistory',
    MAKE_TOKEN_DEPOSIT: 'makeTokenDeposit',
    GET_PROJECT_USAGE_AND_CHARGES: 'getProjectUsageAndCharges',
    GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP: 'getProjectUsageAndChargesCurrentRollup',
    GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP: 'getProjectUsageAndChargesPreviousRollup',
    APPLY_COUPON_CODE: 'applyCouponCode',
    GET_COUPON: `getCoupon`,
};

const {
    SET_BALANCE,
    SET_CREDIT_CARDS,
    SET_DATE,
    CLEAR,
    UPDATE_CARDS_SELECTION,
    UPDATE_CARDS_DEFAULT,
    SET_PAYMENTS_HISTORY,
    SET_PROJECT_USAGE_AND_CHARGES,
    SET_PRICE_SUMMARY,
    SET_PRICE_SUMMARY_FOR_SELECTED_PROJECT,
    SET_COUPON,
} = PAYMENTS_MUTATIONS;

const {
    GET_BALANCE,
    SETUP_ACCOUNT,
    GET_CREDIT_CARDS,
    ADD_CREDIT_CARD,
    TOGGLE_CARD_SELECTION,
    CLEAR_CARDS_SELECTION,
    CLEAR_PAYMENT_INFO,
    MAKE_CARD_DEFAULT,
    REMOVE_CARD,
    GET_PAYMENTS_HISTORY,
    MAKE_TOKEN_DEPOSIT,
    GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP,
    GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP,
    APPLY_COUPON_CODE,
    GET_COUPON,
} = PAYMENTS_ACTIONS;

export class PaymentsState {
    /**
     * balance stores in cents
     */
    public balance: AccountBalance = new AccountBalance();
    public creditCards: CreditCard[] = [];
    public paymentsHistory: PaymentsHistoryItem[] = [];
    public usageAndCharges: ProjectUsageAndCharges[] = [];
    public priceSummary = 0;
    public priceSummaryForSelectedProject = 0;
    public startDate: Date = new Date();
    public endDate: Date = new Date();
    public coupon: Coupon | null = null;
}

interface PaymentsContext {
    state: PaymentsState
    commit: (string, ...unknown) => void
    rootGetters: {
        selectedProject: {
            id: string
        }
    }
}

/**
 * creates payments module with all dependencies
 *
 * @param api - payments api
 */
export function makePaymentsModule(api: PaymentsApi): StoreModule<PaymentsState, PaymentsContext> {
    return {
        state: new PaymentsState(),
        mutations: {
            [SET_BALANCE](state: PaymentsState, balance: AccountBalance): void {
                state.balance = balance;
            },
            [SET_CREDIT_CARDS](state: PaymentsState, creditCards: CreditCard[]): void {
                state.creditCards = creditCards;
            },
            [SET_DATE](state: PaymentsState, dateRange: DateRange) {
                state.startDate = dateRange.startDate;
                state.endDate = dateRange.endDate;
            },
            [UPDATE_CARDS_SELECTION](state: PaymentsState, id: string | null): void {
                state.creditCards = state.creditCards.map(card => {
                    if (card.id === id) {
                        card.isSelected = !card.isSelected;

                        return card;
                    }

                    card.isSelected = false;

                    return card;
                });
            },
            [UPDATE_CARDS_DEFAULT](state: PaymentsState, id: string): void {
                state.creditCards = state.creditCards.map(card => {
                    if (card.id === id) {
                        card.isDefault = !card.isDefault;

                        return card;
                    }

                    card.isDefault = false;

                    return card;
                });
            },
            [SET_PAYMENTS_HISTORY](state: PaymentsState, paymentsHistory: PaymentsHistoryItem[]): void {
                state.paymentsHistory = paymentsHistory;
            },
            [SET_PROJECT_USAGE_AND_CHARGES](state: PaymentsState, usageAndCharges: ProjectUsageAndCharges[]): void {
                state.usageAndCharges = usageAndCharges;
            },
            [SET_PRICE_SUMMARY](state: PaymentsState, charges: ProjectUsageAndCharges[]): void {
                if (charges.length === 0) {
                    state.priceSummary = 0;

                    return;
                }

                const usageItemSummaries: number[] = charges.map(item => item.summary());

                state.priceSummary = usageItemSummaries.reduce((accumulator, current) => accumulator + current);
            },
            [SET_PRICE_SUMMARY_FOR_SELECTED_PROJECT](state: PaymentsState, selectedProjectId: string): void {
                let usageAndChargesForSelectedProject: ProjectUsageAndCharges | undefined;
                if (state.usageAndCharges.length) {
                    usageAndChargesForSelectedProject = state.usageAndCharges.find(item => item.projectId === selectedProjectId);
                }

                if (!usageAndChargesForSelectedProject) {
                    state.priceSummaryForSelectedProject = 0;

                    return;
                }

                state.priceSummaryForSelectedProject = usageAndChargesForSelectedProject.summary();
            },
            [SET_COUPON](state: PaymentsState, coupon: Coupon): void {
                state.coupon = coupon;
            },
            [CLEAR](state: PaymentsState) {
                state.balance = new AccountBalance();
                state.paymentsHistory = [];
                state.usageAndCharges = [];
                state.priceSummary = 0;
                state.creditCards = [];
                state.startDate = new Date();
                state.endDate = new Date();
            },
        },
        actions: {
            [GET_BALANCE]: async function({commit}: PaymentsContext): Promise<AccountBalance> {
                const balance: AccountBalance = await api.getBalance();

                commit(SET_BALANCE, balance);

                return balance;
            },
            [SETUP_ACCOUNT]: async function(): Promise<string> {
                const couponType = await api.setupAccount();

                return couponType;
            },
            [GET_CREDIT_CARDS]: async function({commit}: PaymentsContext): Promise<CreditCard[]> {
                const creditCards = await api.listCreditCards();

                commit(SET_CREDIT_CARDS, creditCards);

                return creditCards;
            },
            [ADD_CREDIT_CARD]: async function(_context: PaymentsContext, token: string): Promise<void> {
                await api.addCreditCard(token);
            },
            [TOGGLE_CARD_SELECTION]: function({commit}: PaymentsContext, id: string): void {
                commit(UPDATE_CARDS_SELECTION, id);
            },
            [CLEAR_CARDS_SELECTION]: function({commit}: PaymentsContext): void {
                commit(UPDATE_CARDS_SELECTION, null);
            },
            [MAKE_CARD_DEFAULT]: async function({commit}: PaymentsContext, id: string): Promise<void> {
                await api.makeCreditCardDefault(id);

                commit(UPDATE_CARDS_DEFAULT, id);
            },
            [REMOVE_CARD]: async function({commit, state}: PaymentsContext, cardId: string): Promise<void> {
                await api.removeCreditCard(cardId);

                commit(SET_CREDIT_CARDS, state.creditCards.filter(card => card.id !== cardId));
            },
            [CLEAR_PAYMENT_INFO]: function({commit}: PaymentsContext): void {
                commit(CLEAR);
            },
            [GET_PAYMENTS_HISTORY]: async function({commit}: PaymentsContext): Promise<void> {
                const paymentsHistory: PaymentsHistoryItem[] = await api.paymentsHistory();

                commit(SET_PAYMENTS_HISTORY, paymentsHistory);
            },
            [MAKE_TOKEN_DEPOSIT]: async function(_context: PaymentsContext, amount: number): Promise<TokenDeposit> {
                return await api.makeTokenDeposit(amount);
            },
            [GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP]: async function({commit, rootGetters}: PaymentsContext): Promise<void> {
                const now = new Date();
                const endUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate(), now.getUTCHours(), now.getUTCMinutes()));
                const startUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 1, 0, 0));

                const usageAndCharges: ProjectUsageAndCharges[] = await api.projectsUsageAndCharges(startUTC, endUTC);

                commit(SET_DATE, new DateRange(startUTC, endUTC));
                commit(SET_PROJECT_USAGE_AND_CHARGES, usageAndCharges);
                commit(SET_PRICE_SUMMARY, usageAndCharges);
                commit(SET_PRICE_SUMMARY_FOR_SELECTED_PROJECT, rootGetters.selectedProject.id);
            },
            [GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP]: async function({commit}: PaymentsContext): Promise<void> {
                const now = new Date();
                const startUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth() - 1, 1, 0, 0));
                const endUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 0, 23, 59, 59));

                const usageAndCharges: ProjectUsageAndCharges[] = await api.projectsUsageAndCharges(startUTC, endUTC);

                commit(SET_DATE, new DateRange(startUTC, endUTC));
                commit(SET_PROJECT_USAGE_AND_CHARGES, usageAndCharges);
                commit(SET_PRICE_SUMMARY, usageAndCharges);
            },
            [APPLY_COUPON_CODE]: async function({commit}: PaymentsContext, code: string): Promise<void> {
                const coupon = await api.applyCouponCode(code);
                commit(SET_COUPON, coupon);
            },
            [GET_COUPON]: async function({commit}: PaymentsContext): Promise<void> {
                const coupon = await api.getCoupon();
                commit(SET_COUPON, coupon);
            }
        },
        getters: {
            canUserCreateFirstProject: (state: PaymentsState): boolean => {
                return state.balance.sum > 0 || state.creditCards.length > 0;
            },
            isTransactionProcessing: (state: PaymentsState): boolean => {
                return state.paymentsHistory.some((paymentsItem: PaymentsHistoryItem) => {
                    return paymentsItem.amount >= 50 && paymentsItem.type === PaymentsHistoryItemType.Transaction
                        && (paymentsItem.status === PaymentsHistoryItemStatus.Pending
                        || paymentsItem.status === PaymentsHistoryItemStatus.Paid
                        || paymentsItem.status === PaymentsHistoryItemStatus.Completed);
                }) && state.balance.sum === 0;
            },
            isBalancePositive: (state: PaymentsState): boolean => {
                return state.balance.sum > 0;
            },
        },
    };
}
