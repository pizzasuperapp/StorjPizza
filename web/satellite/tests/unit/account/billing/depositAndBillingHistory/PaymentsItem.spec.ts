// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import PaymentsItem from '@/components/account/billing/depositAndBillingHistory/PaymentsItem.vue';

import { PaymentsHistoryItem, PaymentsHistoryItemType } from '@/types/payments';
import { createLocalVue, mount } from '@vue/test-utils';

const localVue = createLocalVue();
const itemInvoice = new PaymentsHistoryItem('testId', 'Invoice', 500, 500, 'test', 'test', new Date(1), new Date(1), PaymentsHistoryItemType.Invoice);
const itemCharge = new PaymentsHistoryItem('testId', 'Charge', 500, 500, 'test', 'test', new Date(1), new Date(1), PaymentsHistoryItemType.Charge);
const itemTransaction = new PaymentsHistoryItem('testId', 'Transaction', 500, 500, 'test', 'test', new Date(1), new Date(1), PaymentsHistoryItemType.Transaction);

describe('PaymentsItem', (): void => {
    it('renders correctly if invoice', (): void => {
        const wrapper = mount(PaymentsItem, {
            localVue,
            propsData: {
                billingItem: itemInvoice,
            },
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly if charge', (): void => {
        const wrapper = mount(PaymentsItem, {
            localVue,
            propsData: {
                billingItem: itemCharge,
            },
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly if transaction', (): void => {
        const wrapper = mount(PaymentsItem, {
            localVue,
            propsData: {
                billingItem: itemTransaction,
            },
        });

        expect(wrapper).toMatchSnapshot();
    });
});
