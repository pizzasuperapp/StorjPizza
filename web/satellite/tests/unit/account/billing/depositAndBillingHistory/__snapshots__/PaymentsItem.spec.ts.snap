// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentsItem renders correctly if charge 1`] = `
<div class="container">
  <div class="countdown-container container__item date">
    <div>January 01, 1970</div>
  </div>
  <p class="container__item description">Charge</p>
  <p class="container__item status">Test</p>
  <p class="container__item amount"><b>
      USD $
      <span>
                5.00
            </span></b>
    <!---->
  </p>
  <p class="container__item download"><a target="_blank" href="test" class="download-link">Invoice PDF</a></p>
</div>
`;

exports[`PaymentsItem renders correctly if invoice 1`] = `
<div class="container">
  <div class="countdown-container container__item date">
    <div>January 01, 1970</div>
  </div>
  <p class="container__item description">Invoice</p>
  <p class="container__item status">Test</p>
  <p class="container__item amount"><b>
      USD $
      <span>
                5.00
            </span></b>
    <!---->
  </p>
  <p class="container__item download"><a target="_blank" href="test" class="download-link">Invoice PDF</a></p>
</div>
`;

exports[`PaymentsItem renders correctly if transaction 1`] = `
<div class="container">
  <div class="countdown-container container__item date">
    <div>January 01, 1970</div>
  </div>
  <p class="container__item description">Transaction</p>
  <p class="container__item status">Test</p>
  <p class="container__item amount"><b>
      USD $
      <span>
                5.00
            </span></b> <span>
            of <b>5.00</b></span></p>
  <p class="container__item download"><a target="_blank" href="test" class="download-link">Checkout</a></p>
</div>
`;
